lib/scala/pkg/src/main/scala/org/enso/pkg/ComponentGroup.scala

```diff
@@ -0,0 +1,316 @@
+package org.enso.pkg
+
+import io.circe._
+import io.circe.syntax._
+
+/** The description of component groups provided by the package.
+  *
+  * @param newGroups the list of component groups provided by the package
+  * @param extendedGroups the list of component groups that this package extends
+  */
+case class ComponentGroups(
+  newGroups: List[ComponentGroup],
+  extendedGroups: List[ExtendedComponentGroup]
+)
+object ComponentGroups {
+
+  /** An empty component groups. */
+  val empty: ComponentGroups =
+    ComponentGroups(List(), List())
+
+  /** Fields for use when serializing the [[ComponentGroups]]. */
+  private object Fields {
+    val New     = "new"
+    val Extends = "extends"
+  }
+
+  /** Does the provided JSON object have unknown fields. */
+  private def hasUnknownKeys(cursor: HCursor): Boolean =
+    (ConfigCodecs.objectKeys(cursor) - Fields.New - Fields.Extends).nonEmpty
+
+  /** [[Encoder]] instance for the [[ComponentGroups]]. */
+  implicit val encoder: Encoder[ComponentGroups] = { componentGroups =>
+    val newGroups = Option.unless(componentGroups.newGroups.isEmpty)(
+      Fields.New -> componentGroups.newGroups.asJson
+    )
+    val extendsGroups = Option.unless(componentGroups.extendedGroups.isEmpty)(
+      Fields.Extends -> componentGroups.extendedGroups.asJson
+    )
+    Json.obj(newGroups.toSeq ++ extendsGroups.toSeq: _*)
+  }
+
+  /** [[Decoder]] instance for the [[ComponentGroups]]. */
+  implicit val decoder: Decoder[ComponentGroups] = { json =>
+    def decodeComponentGroups: Either[DecodingFailure, ComponentGroups] = for {
+      newGroups <- json.getOrElse[List[ComponentGroup]](Fields.New)(List())
+      extendsGroups <-
+        json.getOrElse[List[ExtendedComponentGroup]](Fields.Extends)(List())
+    } yield ComponentGroups(newGroups, extendsGroups)
+
+    if (hasUnknownKeys(json)) {
+      Left(
+        DecodingFailure(
+          "Invalid component groups. Valid keys are 'new' or 'extends'.",
+          json.history
+        )
+      )
```

Comment:
Won't this hurt forwards compatibility?

Note that the `package.yaml` is the place things like `launcher` or `project-manager` will read to find the engine version to use. The version of the `launcher` or `project-manager` can often be older than the engine versions that it can support (as they are able to download new engine versions and should be forced to upgrade only when absolutely necessary).

So the effect of adding such an error is that if we ever want to add any new field to `component-groups`, projects using such a field will be unable to be loaded using older launchers/PMs.

Maybe this would be OK if when loading the config, at least by default, we ignored non-fatal errors like this (for example making the `componentGroups` field of Config an `Either[DecodingError, ComponentGroups]` so that even if this sub-config fails to load, other parts of the config can be loaded (since what the launcher/PM usually only needs is the `edition`/`engine-version` fields and nothing more). Then we can afford such forwards-incompatible checks (which have advantages as they will often detect if the user for example made a typo).